test/module-hooks/test-module-hooks-load-url-change.js

```diff
@@ -0,0 +1,29 @@
+'use strict';
+
+const common = require('../common');
+const assert = require('assert');
+const { registerHooks } = require('module');
+const fixtures = require('../common/fixtures');
+
+// This tests shows the url parameter in `load` can be changed in the `nextLoad` call
+
+const hook = registerHooks({
+  resolve(specifier, context, nextLoad) {
+    assert.strictEqual(specifier, 'rfs');
+    return {
+      url: 'file:///rfs',
+      shortCircuit: true,
+    };
+  },
+  load: common.mustCall(function load(url, context, nextLoad) {
+    assert.strictEqual(url, 'file:///rfs');
+    return nextLoad(
+      fixtures.fileURL('module-hooks', `redirected-fs.js`).href,
+      context
+    );
+  }),
+});
+
+assert.strictEqual(require('rfs').exports_for_test, 'redirected fs');
+
+hook.deregister();
```

Comment:
This test is a bit difficult to follow, specifically, how `rfs.exports_for_test` gets set to 'redirected fs'—I presume that's all that module does, but that makes it unnecessarily distanced from the test, for, IMO, no extra value.

I think you don't need any of that magic, nor anything external

```suggestion
const directedUrl = 'redirected-foo';

function nextLoad(url) {
  assert.strictEqual(url, directedUrl);

  return {
    format: 'module',
    shortCircuit: true,
    source: '',
  };
};

function resolveToFakeURL(specifier) {
  assert.strictEqual(specifier, 'foo');

  return {
    url: 'file:///foo',
    shortCircuit: true,
  };
}

function loadSubstitutedURL(url, context, nextLoad) {
  assert.strictEqual(url, 'file:///foo');

  return nextLoad(directedUrl, context);
}

const hooks1 = registerHooks({ load: common.mustCall(nextLoad) });

const hooks2 = registerHooks({
  resolve: common.mustCall(resolveToFakeURL),
  load: common.mustCall(loadSubstitutedURL),
});

hooks1.deregister();
hooks2.deregister();
```

Reply:
Thanks! I agree with your point. However, your test case will pass even without the fix (I added `require` for this). 

From what I understand, the issue is happening in the final stage, where the file is actually being read.

I can also add a module to redirect, which would let us avoid using redirect-fs. It feels like a bit of a trade-off—making the test more readable but adding one more file to the fixtures. What’s your take on that? Also, I’m adding more mockery names in other places.

Reply:
Crap, sorry. Yes, you'll need to add a `require('foo')`. I believe the test will actually not pass without it.

Reply:
Yes, no problem.  I tried with `require`, and it started passing, but it's also passing without my fix. So, the external module is needed for the check since it looks like the function with the problem is called only for the final file read.

Reply:
Okay! Thanks for checking 🙂